Components/LearningCore/Test/gov/sandia/cognition/learning/algorithm/clustering/OptimizedAffinityPropagationTest.java

```diff
@@ -0,0 +1,284 @@
+        assertEquals( selfDivergence, instance.getSelfDivergence() );
+        assertEquals( dampingFactor, instance.getDampingFactor() );
+        assertEquals( OptimizedAffinityPropagation.DEFAULT_MAX_ITERATIONS,
+            instance.getMaxIterations() );
+
+        int maxIterations = 4774;
+        instance = new OptimizedAffinityPropagation<>(metric,
+            selfDivergence, dampingFactor, maxIterations);
+        assertSame( metric, instance.getDivergence() );
+        assertEquals( selfDivergence, instance.getSelfDivergence() );
+        assertEquals( dampingFactor, instance.getDampingFactor() );
+        assertEquals( maxIterations, instance.getMaxIterations() );
+    }
+
+    public void testLearn()
+    {
+        Vector[] data = new Vector[]{
+            new Vector2( -2.341500, 3.696800 ),
+            new Vector2( -1.109200, 3.111700 ),
+            new Vector2( -1.566900, 1.835100 ),
+            new Vector2( -2.658500, 0.664900 ),
+            new Vector2( -4.031700, 2.845700 ),
+            new Vector2( -3.081000, 2.101100 ),
+            new Vector2( 2.588000, 1.781900 ),
+            new Vector2( 3.292300, 3.058500 ),
+            new Vector2( 4.031700, 1.622300 ),
+            new Vector2( 3.081000, -0.611700 ),
+            new Vector2( 0.264100, 0.398900 ),
+            new Vector2( 1.320400, 2.207400 ),
+            new Vector2( 0.193700, 3.643600 ),
+            new Vector2( 1.954200, -0.505300 ),
+            new Vector2( 1.637300, 1.409600 ),
+            new Vector2( -0.123200, -1.516000 ),
+            new Vector2( -1.355600, -3.058500 ),
+            new Vector2( 0.017600, -4.016000 ),
+            new Vector2( 1.003500, -3.590400 ),
+            new Vector2( 0.017600, -2.420200 ),
+            new Vector2( -1.531700, -0.930900 ),
+            new Vector2( -1.144400, 0.505300 ),
+            new Vector2( 0.616200, -1.516000 ),
+            new Vector2( 1.707700, -2.207400 ),
+            new Vector2( 2.095100, 3.430900 )
+        };
+
+        ArrayList<Vector> exemplars = new ArrayList<Vector>();
+        exemplars.add( data[2] );
+        exemplars.add( data[6] );
+        exemplars.add( data[19] );
+
+        int[] membershipCounts = new int[]{9, 8, 8};
+        double selfDivergence = 15.561256;
+
+        OptimizedAffinityPropagation<Vectorizable> instance =
+            new OptimizedAffinityPropagation<Vectorizable>(
+            EuclideanDistanceSquaredMetric.INSTANCE, selfDivergence,
+            0.5, 100 );
+
+        Collection<CentroidCluster<Vectorizable>> clusters =
+            instance.learn( Arrays.asList( data ) );
+
+        assertEquals( 3, clusters.size() );
+
+        for (CentroidCluster<Vectorizable> cluster : clusters)
+        {
+            int index = exemplars.indexOf( cluster.getCentroid() );
+            assertTrue( index >= 0 );
+            assertEquals( membershipCounts[index], cluster.getMembers().size() );
+        }
+    }
```

Comment:
Per Jeremy's comment, it would be good to have a unit test that generates some random data and then checks that the two implementations return the same result.